### ISSUE TITLE
Gallery file organization

### ISSUE THREAD

**Issue by emackey** [adi2412/cesium#1]
We chatted about this in email, but I thought I would try to capture the thoughts as an issue here.

The "old" Sandcastle keeps its gallery demos in a folder with a single HTML file per demo, and CSS and JS embedded in that file.  The build script automatically creates an index of such files, and the gallery loads them from there.  There is a "save as" button in Sandcastle that creates a new file to add to that same folder, that can be indexed and added to the gallery.

The new system breaks these demos out into numbered folders, but there are issues with that.  Using sequential numbers is the wrong solution because (a) the numbers are meaningless to folks who look for demos on the filesystem, and (b) we can't delete obsolete demos without leaving holes in the numbering scheme.

So, we need to:
- [x] Get rid of the sequential numbers entirely, only use demo names for file/folder names.
- [x] Make sure that demo names with special characters are sanitized, for example putting a slash in a demo name should not break the system by attempting to use a slash in a filename.
- [x] Consider the pros and cons of saving a single file vs. saving separate HTML, JS, CSS files along with data/CZML file(s).  Separate files are more readable, but the "Save as" feature gets cumbersome if you have to save more than one.  The "Save as" feature should continue to be able to produce the correct format for inclusion in the gallery.
- [x] If we decide to keep the concept of separate files, the "html" file should be a full functioning HTML document, not just the snippet from the HTML tab.
- [x] The build indexer should automatically build an index of the new gallery.  (If we keep the old single-file-per-demo system, then we can keep the old indexer too, otherwise we have to modify it to understand the new multi-file system).
- [x] The built index should be ignored by `.gitignore`.  Currently it is included in the repo, so it will need to be deleted before it is ignored.

**Comment by adi2412** [comment]
I've decided to stick to the old system and modified my code to read the standalone HTML file for this purpose. I've removed the folder system entirely after comments on the discussion thread. However, I'm thinking of keeping the gallery js file like I had developed, so I'll need to modify the build system for that.

**Comment by adi2412** [comment]
I'm a bit confused on how to modify the build system for this. The current build system runs a script parses through all the demo files and saves the name and the last modified date. I want to instead store JS Objects for each demo containing demo name and description and store this in a dictionary with the key as the demo name. I'm guessing I'll need to run a DOM parser in the build indexer to fetch the description from the file. Is that correct?

**Comment by adi2412** [comment]
I tried modifying the indexer but I'm stuck at one point. The index file that I have created contains a javascript object with the demo names as keys. I can't create new keys since Rhino implements JS 1.7 and adding new keys was only added as part of 1.8.5. I'm not sure how I can overcome this. I had read that there was a plan to move the Cesium build tool to node. Is that going to happen any time soon?

**Comment by emackey** [comment]
Hi @adi2412, unfortunately the Node-based build won't happen anytime soon, but there may still be ways to fix this.  One option is to switch back to an array of objects instead of the large object with names as keys.  You can then search for demos by name like this:

```
var gallery_names = gallery.map(function(d) { return d.name; });
function getDemo(name) {
    var index = gallery_names.indexOf(name);
    return (index < 0) ? undefined : gallery[index];
}
```

**Comment by adi2412** [comment]
Yea, that probably would be the way to go. I'll make changes to my code to fetch the demo this way.

**Comment by adi2412** [comment]
I'd still prefer that the labels and description are part of the index file itself, because it seems unnecessary to read the files when the gallery is loaded. I think we should probably be adding labels and description to the index file.

**Comment by emackey** [comment]
Yeah I'd agree with that.  The code above should work even with the expanded information included in objects in the array.  Let me know if that doesn't make sense.

The original thinking behind the smaller index is a long story.  Basically, Cesium has a "quick" build step that automatically happens anytime you click save in Eclipse, and we want this to stay fast (under a second on modern hardware is best).  Among other things this quick build step converts all our GLSL files to JS string files, and it builds the index of the old Sandcastle gallery.  Originally I had avoided having the indexer go through and pull out the labels & descriptions from all the files, on the theory that keeping this part of the build as fast as possible meant I couldn't peek into all those files during this step.  But I don't think we ever performance-tested this.

This build speedup came at an expense, as you know: The remote client was then responsible for basically downloading all the demos at launch, just to get started.  So I now think that your proposed tradeoff is a good one: the build step looks in all the gallery files and compiles a more complete index, so the client can just download the index and not the whole gallery.  This makes the build step a bit longer, but is a huge win for end users.

If the build gets too long, we have the option of moving this operation out of the "quick" portion of the build, lumping it in with "combine" or some other long-running step.  But I'd prefer not to take that action just yet, ideally the build step can be quick enough to stay were it is, even with the inclusion of labels and descriptions.

**Comment by adi2412** [comment]
That makes sense. I've made changes to the script that generates the index file in de82e5d0f62f46a7e143513c476dccd5d84facf8. All that remains is to change the build.xml to generate the new index file in the Sandcastle-new directory.

**Comment by adi2412** [comment]
And the index file has been added to gitignore already so that is done as well.

**Comment by emackey** [comment]
OK.  Maybe it's time to remove the old Sandcastle and have the new one take over the old one's folder.

**Comment by emackey** [comment]
So I'm confused by what I'm seeing.  It looks like the "sandcastle-new/gallery" folder still has an index file that's checked into Git (not ignored) and not automatically built.  The build procedure does build a new-format index somewhere, maybe in the old Sandcastle gallery folder?

Ideally I'd like to get this branch in good enough shape to open a pull request before the end of GSoC, if possible.  The old Sandcastle would need to be fully removed and replaced with the new one in such a pull request.

**Comment by adi2412** [comment]
Yea, I didn't want to change the build xml file to create the new index file in the sandcastle-new folder. I was first intending to remove the old sandcastle and rename the new folder. The build procedure would continue building it in the old folder itself. I'll push that commit up.
